internal/driver/driver.go

```diff
@@ -467,16 +381,79 @@ func (d *Driver) fetchCommandResponse(cmdUuid string) (string, bool) {
 }
 
 func (d *Driver) AddDevice(deviceName string, protocols map[string]models.ProtocolProperties, adminState models.AdminState) error {
-	d.Logger.Debug(fmt.Sprintf("Device %s is added", deviceName))
+	d.Logger.Debugf("Device %s is added", deviceName)
 	return nil
 }
 
 func (d *Driver) UpdateDevice(deviceName string, protocols map[string]models.ProtocolProperties, adminState models.AdminState) error {
-	d.Logger.Debug(fmt.Sprintf("Device %s is updated", deviceName))
+	d.Logger.Debugf("Device %s is updated", deviceName)
 	return nil
 }
 
 func (d *Driver) RemoveDevice(deviceName string, protocols map[string]models.ProtocolProperties) error {
-	d.Logger.Debug(fmt.Sprintf("Device %s is removed", deviceName))
+	d.Logger.Debugf("Device %s is removed", deviceName)
 	return nil
 }
+
+func createMqttClient(serviceConfig *ServiceConfig) (mqtt.Client, errors.EdgeX) {
+	var scheme = serviceConfig.MQTTBrokerInfo.Schema
+	var brokerUrl = serviceConfig.MQTTBrokerInfo.Host
+	var brokerPort = serviceConfig.MQTTBrokerInfo.Port
+	var authMode = serviceConfig.MQTTBrokerInfo.AuthMode
+	var secretPath = serviceConfig.MQTTBrokerInfo.CredentialsPath
+	var mqttClientId = serviceConfig.MQTTBrokerInfo.ClientId
+	var keepAlive = serviceConfig.MQTTBrokerInfo.KeepAlive
+
+	uri := &url.URL{
+		Scheme: strings.ToLower(scheme),
+		Host:   fmt.Sprintf("%s:%d", brokerUrl, brokerPort),
+	}
+
+	err := SetCredentials(uri, "init", authMode, secretPath)
+	if err != nil {
+		return nil, errors.NewCommonEdgeXWrapper(err)
+	}
+
+	var client mqtt.Client
+	for i := 1; i <= serviceConfig.MQTTBrokerInfo.ConnEstablishingRetry; i++ {
+		client, err = mqttClient(mqttClientId, uri, keepAlive)
+		if err != nil && i == serviceConfig.MQTTBrokerInfo.ConnEstablishingRetry {
+			return nil, errors.NewCommonEdgeXWrapper(err)
+		} else if err != nil {
+			driver.Logger.Errorf("Fail to initial the MQTT conn, %v ", err)
+			time.Sleep(time.Duration(serviceConfig.MQTTBrokerInfo.ConnEstablishingRetry) * time.Second)
+			driver.Logger.Warn("Retry to initial the MQTT conn")
+			continue
+		}
+		break
+	}
+	return client, nil
+}
+
+func mqttClient(clientID string, uri *url.URL, keepAlive int) (mqtt.Client, error) {
+	driver.Logger.Infof("Create MQTT client and connection: uri=%v clientID=%v ", uri.String(), clientID)
+	opts := mqtt.NewClientOptions()
+	opts.AddBroker(fmt.Sprintf("%s://%s", uri.Scheme, uri.Host))
+	opts.SetClientID(clientID)
+	opts.SetUsername(uri.User.Username())
+	password, _ := uri.User.Password()
+	opts.SetPassword(password)
+	opts.SetKeepAlive(time.Second * time.Duration(keepAlive))
+	opts.SetConnectionLostHandler(func(client mqtt.Client, e error) {
+		driver.Logger.Warnf("Connection lost : %v", e)
+		token := client.Connect()
+		if token.Wait() && token.Error() != nil {
+			driver.Logger.Warnf("Reconnection failed : %v", token.Error())
+		} else {
+			driver.Logger.Warn("Reconnection sucessful")
+		}
+	})
+
```

Comment:
We do what to set the `opts.OnConnect` handler and do the subscribes to Incoming and Response topics there so if connection is lost it will re-subscribe.
Example here for App services: https://github.com/edgexfoundry/app-functions-sdk-go/blob/master/internal/trigger/mqtt/mqtt.go#L131

Reply:
Added OnConnect handler and verified it works as you suggested.